packages/network/src/logic/proxy/ProxyStreamConnectionClient.ts

```diff
@@ -0,0 +1,358 @@
+import { TrackerManager } from '../TrackerManager'
+import { StreamPartManager } from '../StreamPartManager'
+import { NodeToNode } from '../../protocol/NodeToNode'
+import { NodeId } from '../../identifiers'
+import { Node, Event as NodeEvent } from '../Node'
+import {
+    ProxyConnectionResponse,
+    ProxyDirection,
+    StreamPartID,
+    UnsubscribeRequest
+} from '@streamr/protocol'
+import { Logger, wait, withTimeout } from "@streamr/utils"
+import { Propagation } from '../propagation/Propagation'
+import { shuffle } from 'lodash'
+import { EventEmitter } from "events"
+
+const logger = new Logger(module)
+
+export interface ProxyStreamConnectionClientOptions {
+    trackerManager: TrackerManager
+    streamPartManager: StreamPartManager
+    nodeToNode: NodeToNode
+    propagation: Propagation
+    node: Node
+    nodeConnectTimeout: number
+}
+
+enum State {
+    NEGOTIATING,
+    ACCEPTED,
+}
+
+interface ProxyCandidate {
+    state?: State
+    direction: ProxyDirection
+    userId: string
+}
+
+interface ProxyTargets {
```

Comment:
What the `target` word means here? Or something like ProxyDefinition, if it defined which nodes we use as proxies for a stream? If it is just a definition, then `numOfTargets` can be `connectionCount` as in the public API?

What's the correlation between `connections` map, or in `ProxyCandidate.state`: if a new connection is made, do we need to update both?